packages/playwright-core/src/server/injected/webSocketMock.ts

```diff
@@ -0,0 +1,336 @@
+    s = globalThis.atob(s);
+    const b = new Uint8Array(s.length);
+    for (let i = 0; i < s.length; i++)
+      b[i] = s.charCodeAt(i);
+    return b.buffer;
+  }
+
+  // Note: this function tries to be synchronous when it can to preserve the ability to send
+  // multiple messages synchronously in the same order and then synchronously close.
+  function messageToData(message: WebSocketMessage, cb: (data: WSData) => any) {
+    if (message instanceof globalThis.Blob)
+      return message.arrayBuffer().then(buffer => cb(bufferToData(new Uint8Array(buffer))));
+    if (typeof message === 'string')
+      return cb({ data: message, isBase64: false });
+    if (ArrayBuffer.isView(message))
+      return cb(bufferToData(new Uint8Array(message.buffer, message.byteOffset, message.byteLength)));
+    return cb(bufferToData(new Uint8Array(message)));
+  }
+
+  function dataToMessage(data: WSData, binaryType: 'blob' | 'arraybuffer'): WebSocketMessage {
+    if (!data.isBase64)
+      return data.data;
+    const buffer = stringToBuffer(data.data);
+    return binaryType === 'arraybuffer' ? buffer : new Blob([buffer]);
+  }
+
+  const binding = (globalThis as any).__pwWebSocketBinding as (message: BindingPayload) => void;
+  const NativeWebSocket: typeof WebSocket = globalThis.WebSocket;
+  const idToWebSocket = new Map<string, WebSocketMock>();
+  (globalThis as any).__pwWebSocketDispatch = (request: APIRequest) => {
+    const ws = idToWebSocket.get(request.id);
+    if (!ws)
+      return;
+    if (request.type === 'connect')
+      ws._apiConnect(request.fuse);
+    if (request.type === 'ensureOpened')
+      ws._apiEnsureOpened();
+    if (request.type === 'sendToPage')
+      ws._apiSendToPage(dataToMessage(request.data, ws.binaryType));
+    if (request.type === 'close')
+      ws._apiClose(request.code, request.reason, request.wasClean);
+    if (request.type === 'sendToServer')
+      ws._apiSendToServer(dataToMessage(request.data, ws.binaryType));
+  };
+
+  class WebSocketMock extends EventTarget {
+    static readonly CONNECTING: 0 = 0; // WebSocket.CONNECTING
+    static readonly OPEN: 1 = 1; // WebSocket.OPEN
+    static readonly CLOSING: 2 = 2; // WebSocket.CLOSING
+    static readonly CLOSED: 3 = 3; // WebSocket.CLOSED
+
+    CONNECTING: 0 = 0; // WebSocket.CONNECTING
+    OPEN: 1 = 1; // WebSocket.OPEN
+    CLOSING: 2 = 2; // WebSocket.CLOSING
+    CLOSED: 3 = 3; // WebSocket.CLOSED
+
+    private _oncloseListener: WebSocket['onclose'] = null;
+    private _onerrorListener: WebSocket['onerror'] = null;
+    private _onmessageListener: WebSocket['onmessage'] = null;
+    private _onopenListener: WebSocket['onopen'] = null;
+
+    bufferedAmount: number = 0;
+    extensions: string = '';
+    protocol: string = '';
+    readyState: number = 0;
+    readonly url: string;
+
+    private _id: string;
+    private _origin: string = '';
+    private _protocols?: string | string[];
+    private _ws?: WebSocket;
+    private _fused = false;
```

Comment:
for me fuse/fused isn't idomatic, maybe 'intercepted' or 'mocked' is a better variable name?